### PR TITLE
Implement strict UTF-8 validation

### LINKED ISSUE #2412: UTF8 validation is too sloppy
Some invalid utf8 sequences are allowed by the utf8 validator code.

https://stackoverflow.com/questions/6555015/check-for-invalid-utf8

https://github.com/ethereum/solidity/pull/2386

### PR DESCRIPTION
**Pull request by axic**
Fixes #2412.

### CODE REVIEW

**Comment by axic** [comment]
I think this is now super strict. Someone needs to add tests for all these cases... 😉

File: libdevcore/UTF8.cpp [hunk 1]
```diff
@@ -27,6 +27,50 @@
...
 
+namespace
+{
+
+/// Validate byte sequence against Unicode chapter 3 Table 3-7.
+bool isWellFormed(unsigned char byte1, unsigned char byte2)
+{
+	switch (byte1)
+	{
+	case 0xc0 ... 0xc1:
```
Comment (chriseth) [inline libdevcore/UTF8.cpp]: gcc is complaining here that `0xc1` is beyond the value range of `unsigned char`. Perhaps make it `unsigned`?

**Review by chriseth — APPROVED** [review]

File: libdevcore/UTF8.cpp (hunk 1 above)
Comment (axic) [reply libdevcore/UTF8.cpp]: Fixed.

**Comment by axic** [comment]
Unfortunately range cases are not supported by MSVC so this needs to be reworked.

**Comment by axic** [comment]
I've liked the case ranges way better..

**Comment by chriseth** [comment]
This is C++, you could do something like

```
Switch(byte1)
.case(0x12, 0x22, [&]{
})
.case(0x44, 0x55, [&]{
});
```
:-)

File: libdevcore/UTF8.cpp [hunk 2]
```diff
@@ -77,5 +133,11 @@ bool validateUTF8(std::string const& _input, size_t& _invalidPosition)
 	return false;
 }
 
+}
+
+bool validateUTF8(std::string const& _input, size_t& _invalidPosition)
+{
+	return validateUTF8((const unsigned char *)_input.c_str(), _input.length(), _invalidPosition);
```
Comment (chriseth) [inline libdevcore/UTF8.cpp]: Please use `reinterpret_cast<unsigned char const*>(...)`.

Comment (axic) [reply libdevcore/UTF8.cpp]: Fixed

**Comment by axic** [comment]
Adding tests right now.

**Comment by axic** [comment]
@chriseth added comprehensive tests.

**Pull request merged**